File: android/app/src/main/java/com/cras/app/domain/TemporalService.kt
package com.cras.app.domain

import com.cras.app.models.Plan
import com.cras.app.models.Task
import java.time.Instant
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime
import java.time.ZoneId
import java.time.ZoneOffset
import java.time.ZonedDateTime
import java.time.format.DateTimeFormatter
import java.util.Locale

enum class TimedPlanType(val value: String) {
    INSTANT("instant"),
    FLOATING("floating");

    companion object {
        fun fromValue(value: String?): TimedPlanType? = when (value?.lowercase()) {
            "instant" -> INSTANT
            "floating" -> FLOATING
            else -> null
        }
    }
}

data class CreatePlanParams(
    val date: String? = null,
    val time: String? = null,
    val type: TimedPlanType? = null,
    val effectiveDefault: TimedPlanType = TimedPlanType.INSTANT,
    val zoneId: ZoneId = ZoneId.systemDefault(),
    val now: Instant = Instant.now()
)

data class PlanDisplayInfo(
    val dateLabel: String,
    val timeLabel: String?,
    val typeLabel: String?,
    val isOverdue: Boolean
)

data class UpcomingDayGroup(
    val date: String,
    val dateLabel: String,
    val tasks: List<Task>
)

data class UpcomingResult(
    val overdue: List<Task>,
    val groups: List<UpcomingDayGroup>
)

private val DATE_FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd")
private val TIME_FORMATTER_SHORT = DateTimeFormatter.ofPattern("HH:mm")
private val FRIENDLY_DATE_FORMATTER = DateTimeFormatter.ofPattern("EEE, MMM d", Locale.ENGLISH)

/**
 * Returns the device-local calendar date formatted as YYYY-MM-DD.
 */
fun getDeviceLocalDate(
    now: Instant = Instant.now(),
    zoneId: ZoneId = ZoneId.systemDefault()
): String {
    val zonedDateTime = now.atZone(zoneId)
    return zonedDateTime.toLocalDate().format(DATE_FORMATTER)
}

/**
 * Formats a calendar date string (YYYY-MM-DD) into a friendly label ("Today", "Tomorrow", "Yesterday", "Wed, Aug 19").
 */
fun formatFriendlyDateLabel(
    dateStr: String,
    todayStr: String,
    tomorrowStr: String,
    yesterdayStr: String? = null
): String {
    if (dateStr == todayStr) {
        return "Today"
    }
    if (dateStr == tomorrowStr) {
        return "Tomorrow"
    }
    if (yesterdayStr != null && dateStr == yesterdayStr) {
        return "Yesterday"
    }

    return try {
        val parsedDate = LocalDate.parse(dateStr, DATE_FORMATTER)
        parsedDate.format(FRIENDLY_DATE_FORMATTER)
    } catch (_: Exception) {
        dateStr
    }
}

/**
 * Extracts the device-local calendar date (YYYY-MM-DD) representing a task plan.
 * For Date-only and Floating, this is the stored calendar date.
 * For Instant, this is the device-local date of the UTC moment in the acting zoneId.
 */
fun getPlanLocalDate(
    plan: Plan?,
    zoneId: ZoneId = ZoneId.systemDefault()
): String? {
    if (plan == null) return null

    return when (plan) {
        is Plan.DateOnly -> plan.date
        is Plan.Floating -> plan.date
        is Plan.Instant -> {
            try {
                val instant = Instant.parse(plan.at)
                val zonedDateTime = instant.atZone(zoneId)
                zonedDateTime.toLocalDate().format(DATE_FORMATTER)
            } catch (_: Exception) {
                null
            }
        }
    }
}

/**
 * Creates a contract-compliant Plan object from user inputs.
 * - If date is omitted or blank, returns null.
 * - If time is omitted or blank, returns Date-only plan Plan.DateOnly(date) without fake midnight or type.
 * - If time is provided, resolves explicit type or falls back to effective default.
 * - For Instant, converts the viewing device's local date + local time into RFC 3339 UTC ISO string.
 * - For Floating, stores the exact date and time without timezone adjustment.
 */
fun createPlanFromInputs(params: CreatePlanParams): Plan? {
    val rawDate = params.date?.trim()
    if (rawDate.isNullOrEmpty()) {
        return null
    }

    val rawTime = params.time?.trim()
    if (rawTime.isNullOrEmpty()) {
        return Plan.DateOnly(date = rawDate)
    }

    // Format time as HH:mm
    val timeParts = rawTime.split(":")
    if (timeParts.size < 2) {
        return Plan.DateOnly(date = rawDate)
    }
    val hours = timeParts[0].padStart(2, '0')
    val minutes = timeParts[1].padStart(2, '0')
    val normalizedTime = "$hours:$minutes"

    val resolvedType = params.type ?: params.effectiveDefault

    if (resolvedType == TimedPlanType.FLOATING) {
        return Plan.Floating(
            date = rawDate,
            time = normalizedTime
        )
    }

    // Instant: resolve local date & time in zoneId to UTC ISO string
    return try {
        val parsedDate = LocalDate.parse(rawDate, DATE_FORMATTER)
        val parsedTime = LocalTime.of(hours.toInt(), minutes.toInt())
        val localDateTime = LocalDateTime.of(parsedDate, parsedTime)
        val zonedDateTime = localDateTime.atZone(params.zoneId)
        val utcInstant = zonedDateTime.toInstant()
        Plan.Instant(at = utcInstant.toString())
    } catch (_: Exception) {
        Plan.DateOnly(date = rawDate)
    }
}

/**
 * Checks whether an open task is overdue relative to device's today.
 */
fun isTaskOverdue(
    task: Task,
    now: Instant = Instant.now(),
    zoneId: ZoneId = ZoneId.systemDefault()
): Boolean {
    if (task.completedAt != null || task.plan == null) {
        return false
    }
    val planDay = getPlanLocalDate(task.plan, zoneId) ?: return false
    val todayStr = getDeviceLocalDate(now, zoneId)
    return planDay < todayStr
}

/**
 * Formats a Plan for display in the UI using relative date semantics.
 */
fun formatPlanDisplay(
    plan: Plan?,
    now: Instant = Instant.now(),
    zoneId: ZoneId = ZoneId.systemDefault()
): PlanDisplayInfo? {
    if (plan == null) return null

    val todayLocalDate = now.atZone(zoneId).toLocalDate()
    val todayStr = todayLocalDate.format(DATE_FORMATTER)
    val tomorrowStr = todayLocalDate.plusDays(1).format(DATE_FORMATTER)
    val yesterdayStr = todayLocalDate.minusDays(1).format(DATE_FORMATTER)

    val planDateStr = getPlanLocalDate(plan, zoneId) ?: return null
    val dateLabel = formatFriendlyDateLabel(planDateStr, todayStr, tomorrowStr, yesterdayStr)

    var timeLabel: String? = null
    var typeLabel: String? = null

    when (plan) {
        is Plan.DateOnly -> {
            timeLabel = null
            typeLabel = null
        }
        is Plan.Floating -> {
            typeLabel = "Floating"
            val parts = plan.time.split(":")
            timeLabel = if (parts.size >= 2) "${parts[0].padStart(2, '0')}:${parts[1].padStart(2, '0')}" else plan.time
        }
        is Plan.Instant -> {
            typeLabel = "Instant"
            try {
                val instant = Instant.parse(plan.at)
                val zonedDateTime = instant.atZone(zoneId)
                timeLabel = zonedDateTime.toLocalTime().format(TIME_FORMATTER_SHORT)
            } catch (_: Exception) {
                timeLabel = null
            }
        }
    }

    val isOverdue = planDateStr < todayStr

    return PlanDisplayInfo(
        dateLabel = dateLabel,
        timeLabel = timeLabel,
        typeLabel = typeLabel,
        isOverdue = isOverdue
    )
}

/**
 * Filters tasks belonging to the Today view:
 * "The view of open Tasks whose plan day is today or earlier, using the viewing device's local calendar date.
 * Completed Tasks are not shown."
 */
fun filterTodayTasks(
    tasks: List<Task>,
    now: Instant = Instant.now(),
    zoneId: ZoneId = ZoneId.systemDefault()
): List<Task> {
    val todayStr = getDeviceLocalDate(now, zoneId)

    return tasks
        .filter { task ->
            if (task.completedAt != null || task.plan == null) {
                false
            } else {
                val planDay = getPlanLocalDate(task.plan, zoneId)
                planDay != null && planDay <= todayStr
            }
        }
        .sortedWith(
            compareBy<Task> { getPlanLocalDate(it.plan, zoneId) ?: "" }
                .thenBy { it.priority }
        )
}

/**
 * Filters tasks belonging to the Upcoming view:
 * "The view of open dated Tasks grouped by day, from today into the future, with overdue in a strip at the top.
 * There is no 7-day window."
 */
fun filterUpcomingTasks(
    tasks: List<Task>,
    now: Instant = Instant.now(),
    zoneId: ZoneId = ZoneId.systemDefault()
): UpcomingResult {
    val todayLocalDate = now.atZone(zoneId).toLocalDate()
    val todayStr = todayLocalDate.format(DATE_FORMATTER)
    val tomorrowStr = todayLocalDate.plusDays(1).format(DATE_FORMATTER)

    val openDatedTasks = tasks.filter { it.completedAt == null && it.plan != null }

    val overdue = mutableListOf<Task>()
    val futureTasksByDate = mutableMapOf<String, MutableList<Task>>()

    for (task in openDatedTasks) {
        val planDay = getPlanLocalDate(task.plan, zoneId) ?: continue
        if (planDay < todayStr) {
            overdue.add(task)
        } else {
            val list = futureTasksByDate.getOrPut(planDay) { mutableListOf() }
            list.add(task)
        }
    }

    // Sort overdue tasks by plan date ascending
    overdue.sortWith(
        compareBy<Task> { getPlanLocalDate(it.plan, zoneId) ?: "" }
            .thenBy { it.priority }
    )

    // Sort date groups ascending
    val sortedDates = futureTasksByDate.keys.sorted()
    val groups = sortedDates.map { date ->
        val dateLabel = formatFriendlyDateLabel(date, todayStr, tomorrowStr)
        val groupTasks = futureTasksByDate.getValue(date).sortedBy { it.priority }
        UpcomingDayGroup(
            date = date,
            dateLabel = dateLabel,
            tasks = groupTasks
        )
    }

    return UpcomingResult(
        overdue = overdue,
        groups = groups
    )
}

File: web/src/services/temporalService.ts
import type { Plan, Task } from "../contracts/task";

export type TimedPlanType = "instant" | "floating";

export interface CreatePlanParams {
  readonly date?: string | null;
  readonly time?: string | null;
  readonly type?: TimedPlanType | null;
  readonly effectiveDefault: TimedPlanType;
  readonly now?: Date;
}

export interface PlanDisplayInfo {
  readonly dateLabel: string;
  readonly timeLabel: string | null;
  readonly typeLabel: "Instant" | "Floating" | null;
  readonly isOverdue: boolean;
}

export interface UpcomingDayGroup {
  readonly date: string;
  readonly dateLabel: string;
  readonly tasks: Task[];
}

export interface UpcomingResult {
  readonly overdue: Task[];
  readonly groups: UpcomingDayGroup[];
}

/**
 * Returns the device-local calendar date formatted as YYYY-MM-DD.
 */
export function getDeviceLocalDate(now: Date = new Date()): string {
  const year = now.getFullYear();
  const month = String(now.getMonth() + 1).padStart(2, "0");
  const day = String(now.getDate()).padStart(2, "0");
  return `${year}-${month}-${day}`;
}

/**
 * Extracts the device-local calendar date (YYYY-MM-DD) representing a task plan.
 * For Date-only and Floating, this is the stored calendar date.
 * For Instant, this is the device-local date of the UTC moment.
 */
export function getPlanLocalDate(plan: Plan | null | undefined): string | null {
  if (!plan) return null;

  if ("type" in plan) {
    if (plan.type === "floating") {
      return plan.date;
    }
    if (plan.type === "instant") {
      const instantDate = new Date(plan.at);
      if (Number.isNaN(instantDate.getTime())) {
        return null;
      }
      return getDeviceLocalDate(instantDate);
    }
  }

  if ("date" in plan && typeof plan.date === "string") {
    return plan.date;
  }

  return null;
}

/**
 * Creates a contract-compliant Plan object from user inputs.
 * - If date is omitted, returns null.
 * - If time is omitted, returns Date-only plan `{ date }` without fake midnight or type.
 * - If time is provided, resolves explicit type or falls back to effective default.
 * - For Instant, converts the viewing device's local date + local time into RFC 3339 UTC ISO string.
 * - For Floating, stores the exact date and time without timezone adjustment.
 */
export function createPlanFromInputs(params: CreatePlanParams): Plan {
  const rawDate = params.date?.trim();
  if (!rawDate) {
    return null;
  }

  const rawTime = params.time?.trim();
  if (!rawTime) {
    return { date: rawDate };
  }

  // Format time as HH:mm (or HH:mm:ss)
  const timeParts = rawTime.split(":");
  if (timeParts.length < 2) {
    return { date: rawDate };
  }
  const hours = timeParts[0].padStart(2, "0");
  const minutes = timeParts[1].padStart(2, "0");
  const normalizedTime = `${hours}:${minutes}`;

  const resolvedType = params.type || params.effectiveDefault || "instant";

  if (resolvedType === "floating") {
    return {
      type: "floating",
      date: rawDate,
      time: normalizedTime,
    };
  }

  // Instant: resolve local date & time to UTC ISO string
  const [yearStr, monthStr, dayStr] = rawDate.split("-");
  const year = Number.parseInt(yearStr, 10);
  const month = Number.parseInt(monthStr, 10) - 1;
  const day = Number.parseInt(dayStr, 10);
  const h = Number.parseInt(hours, 10);
  const m = Number.parseInt(minutes, 10);

  const localDateTime = new Date(year, month, day, h, m, 0, 0);
  return {
    type: "instant",
    at: localDateTime.toISOString(),
  };
}

/**
 * Checks whether an open task is overdue relative to device's today.
 */
export function isTaskOverdue(task: Task, now: Date = new Date()): boolean {
  if (task.completedAt !== null || !task.plan) {
    return false;
  }
  const planDay = getPlanLocalDate(task.plan);
  if (!planDay) return false;

  const today = getDeviceLocalDate(now);
  return planDay < today;
}

/**
 * Formats a Plan for display in the UI using relative date semantics.
 */
export function formatPlanDisplay(
  plan: Plan | null | undefined,
  options?: { now?: Date },
): PlanDisplayInfo | null {
  if (!plan) return null;

  const now = options?.now || new Date();
  const todayStr = getDeviceLocalDate(now);

  const yesterdayDate = new Date(
    now.getFullYear(),
    now.getMonth(),
    now.getDate() - 1,
  );
  const yesterdayStr = getDeviceLocalDate(yesterdayDate);

  const tomorrowDate = new Date(
    now.getFullYear(),
    now.getMonth(),
    now.getDate() + 1,
  );
  const tomorrowStr = getDeviceLocalDate(tomorrowDate);

  const planDateStr = getPlanLocalDate(plan);
  if (!planDateStr) return null;

  let dateLabel: string;
  if (planDateStr === todayStr) {
    dateLabel = "Today";
  } else if (planDateStr === tomorrowStr) {
    dateLabel = "Tomorrow";
  } else if (planDateStr === yesterdayStr) {
    dateLabel = "Yesterday";
  } else {
    // Format friendly date, e.g. "Wed, Aug 20" or "Aug 20"
    const [y, m, d] = planDateStr.split("-").map((n) => Number.parseInt(n, 10));
    const targetDate = new Date(y, m - 1, d);
    dateLabel = targetDate.toLocaleDateString(undefined, {
      weekday: "short",
      month: "short",
      day: "numeric",
    });
  }

  let timeLabel: string | null = null;
  let typeLabel: "Instant" | "Floating" | null = null;

  if ("type" in plan) {
    if (plan.type === "floating") {
      typeLabel = "Floating";
      // Take HH:mm
      const parts = plan.time.split(":");
      timeLabel = `${parts[0]}:${parts[1]}`;
    } else if (plan.type === "instant") {
      typeLabel = "Instant";
      const instantDate = new Date(plan.at);
      const h = String(instantDate.getHours()).padStart(2, "0");
      const m = String(instantDate.getMinutes()).padStart(2, "0");
      timeLabel = `${h}:${m}`;
    }
  }

  const isOverdue = planDateStr < todayStr;

  return {
    dateLabel,
    timeLabel,
    typeLabel,
    isOverdue,
  };
}

/**
 * Filters tasks belonging to the Today view:
 * "The view of open Tasks whose plan day is today or earlier, using the viewing device's local calendar date.
 * Completed Tasks are not shown."
 */
export function filterTodayTasks(
  tasks: readonly Task[],
  now: Date = new Date(),
): Task[] {
  const todayStr = getDeviceLocalDate(now);
  return tasks
    .filter((task) => {
      if (task.completedAt !== null || !task.plan) {
        return false;
      }
      const planDay = getPlanLocalDate(task.plan);
      return planDay !== null && planDay <= todayStr;
    })
    .slice()
    .sort((a, b) => {
      const dayA = getPlanLocalDate(a.plan) || "";
      const dayB = getPlanLocalDate(b.plan) || "";
      if (dayA !== dayB) {
        return dayA.localeCompare(dayB);
      }
      return a.priority - b.priority;
    });
}

/**
 * Filters tasks belonging to the Upcoming view:
 * "The view of open dated Tasks grouped by day, from today into the future, with overdue in a strip at the top.
 * There is no 7-day window."
 */
export function filterUpcomingTasks(
  tasks: readonly Task[],
  now: Date = new Date(),
): UpcomingResult {
  const todayStr = getDeviceLocalDate(now);
  const tomorrowDate = new Date(
    now.getFullYear(),
    now.getMonth(),
    now.getDate() + 1,
  );
  const tomorrowStr = getDeviceLocalDate(tomorrowDate);

  const openDatedTasks = tasks.filter(
    (t) => t.completedAt === null && t.plan !== null,
  );

  const overdue: Task[] = [];
  const futureTasksByDate = new Map<string, Task[]>();

  for (const task of openDatedTasks) {
    const planDay = getPlanLocalDate(task.plan);
    if (!planDay) continue;

    if (planDay < todayStr) {
      overdue.push(task);
    } else {
      const existing = futureTasksByDate.get(planDay) || [];
      existing.push(task);
      futureTasksByDate.set(planDay, existing);
    }
  }

  // Sort overdue tasks by date ascending (oldest overdue first)
  overdue.sort((a, b) => {
    const dayA = getPlanLocalDate(a.plan) || "";
    const dayB = getPlanLocalDate(b.plan) || "";
    return dayA.localeCompare(dayB);
  });

  // Sort dates ascending
  const sortedDates = Array.from(futureTasksByDate.keys()).sort();
  const groups: UpcomingDayGroup[] = sortedDates.map((date) => {
    let dateLabel: string;
    if (date === todayStr) {
      dateLabel = "Today";
    } else if (date === tomorrowStr) {
      dateLabel = "Tomorrow";
    } else {
      const [y, m, d] = date.split("-").map((n) => Number.parseInt(n, 10));
      const targetDate = new Date(y, m - 1, d);
      dateLabel = targetDate.toLocaleDateString(undefined, {
        weekday: "short",
        month: "short",
        day: "numeric",
      });
    }

    const groupTasks = futureTasksByDate.get(date) || [];
    groupTasks.sort((a, b) => a.priority - b.priority);

    return {
      date,
      dateLabel,
      tasks: groupTasks,
    };
  });

  return {
    overdue,
    groups,
  };
}
